File: elixir/collatz-conjecture/lib/collatz_conjecture.ex
defmodule CollatzConjecture do
  import Integer, only: [is_even: 1]

  @doc """
  calc/1 takes an integer and returns the number of steps required to get the
  number to 1 when following the rules:
    - if number is odd, multiply with 3 and add 1
    - if number is even, divide by 2
  """

  # @spec calc(1) :: 0
  def calc(1), do: 0

  @spec calc(number) :: steps
        when number: pos_integer(),
             steps: non_neg_integer()
  def calc(number) when is_integer(number) and number > 0 do
    do_calc(number, 0)
  end

  # @spec calc(any()) :: no_return()

  @spec do_calc(number, step) :: steps
        when number: pos_integer(),
             step: non_neg_integer(),
             steps: non_neg_integer()
  def do_calc(1, steps), do: steps

  def do_calc(number, step) when is_even(number) do
    do_calc(div(number, 2), step + 1)
  end

  def do_calc(number, step) do
    do_calc(3 * number + 1, step + 1)
  end
end

File: prolog/collatz-conjecture/collatz_conjecture.pl
:- module(collatz_conjecture, [collatz_steps/2]).

%!  collatz_steps(+N:int, -Steps:int) is det.
%
%   Compute the number of =Steps= required to reach 1.
collatz_steps(1, 0) :- !.
collatz_steps(N, Steps) :-
    must_be(integer, N),
    N > 1,
    (  even(N)
    -> Next is N // 2
    ;  Next is 3 * N + 1
    ),
    collatz_steps(Next, NextSteps),
    Steps is 1 + NextSteps.

%!  even(+N:int) is det.
%   True if =N= is even.
even(N) :-
    0 is N mod 2.
